activerecord/lib/active_record/migration.rb

```diff
@@ -1014,10 +1015,31 @@ def mtime
   end
 
   class MigrationContext #:nodoc:
-    attr_reader :migrations_paths
+    attr_reader :migrations_paths, :schema_migration
 
-    def initialize(migrations_paths)
+    def initialize(migrations_paths, schema_migration)
       @migrations_paths = migrations_paths
+      @schema_migration = schema_migration
+    end
+
+    def table_exists?
+      schema_migration.table_exists?
+    end
+
+    def all_versions
+      schema_migration.all_versions.map(&:to_i)
+    end
+
+    def normalized_versions
+      schema_migration.normalized_versions
+    end
+
+    def normalize_migration_number(version)
+      schema_migration.normalize_migration_number(version)
+    end
+
+    def create_schema_migration_table
```

Comment:
Are these internal methods should be public?
`create_schema_migration_table` is not used yet, and others are not used from public.

https://github.com/rails/rails/compare/9ef5b7edbad2383ae50413ffd3d71fc6ab1e6b3e...remove-internal-methods-in-migration-context

```patch
From 27cf3da4c0ede28302bbb4e27fa53d1aadb6bc7d Mon Sep 17 00:00:00 2001
From: Ryuta Kamizono <kamipo@gmail.com>
Date: Mon, 10 Jun 2019 20:34:03 +0900
Subject: [PATCH] Remove internal methods which are not used from public

---
 .../connection_adapters/abstract_adapter.rb   |  2 +-
 activerecord/lib/active_record/migration.rb   | 30 ++++---------------
 2 files changed, 6 insertions(+), 26 deletions(-)

diff --git a/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb b/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
index 3a08900f3d07..3d0993dfd7e1 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
@@ -166,7 +166,7 @@ def migration_context # :nodoc:
         MigrationContext.new(migrations_paths, schema_migration)
       end
 
-      def schema_migration
+      def schema_migration # :nodoc:
         @schema_migration ||= begin
                                 conn = self
                                 spec_name = conn.pool.spec.name
diff --git a/activerecord/lib/active_record/migration.rb b/activerecord/lib/active_record/migration.rb
index 0256d2b710f9..aa2e7f95aec0 100644
--- a/activerecord/lib/active_record/migration.rb
+++ b/activerecord/lib/active_record/migration.rb
@@ -885,7 +885,7 @@ def method_missing(method, *arguments, &block)
 
     def copy(destination, sources, options = {})
       copied = []
-      schema_migration = options[:schema_migration]
+      schema_migration = ActiveRecord::Base.connection.schema_migration
 
       FileUtils.mkdir_p(destination) unless File.exist?(destination)
 
@@ -1022,26 +1022,6 @@ def initialize(migrations_paths, schema_migration)
       @schema_migration = schema_migration
     end
 
-    def table_exists?
-      schema_migration.table_exists?
-    end
-
-    def all_versions
-      schema_migration.all_versions.map(&:to_i)
-    end
-
-    def normalized_versions
-      schema_migration.normalized_versions
-    end
-
-    def normalize_migration_number(version)
-      schema_migration.normalize_migration_number(version)
-    end
-
-    def create_schema_migration_table
-      schema_migration.create_table
-    end
-
     def migrate(target_version = nil, &block)
       case
       when target_version.nil?
@@ -1092,8 +1072,8 @@ def open
     end
 
     def get_all_versions
-      if table_exists?
-        all_versions
+      if schema_migration.table_exists?
+        schema_migration.all_versions.map(&:to_i)
       else
         []
       end
@@ -1130,12 +1110,12 @@ def migrations
     end
 
     def migrations_status
-      db_list = normalized_versions
+      db_list = schema_migration.normalized_versions
 
       file_list = migration_files.map do |file|
         version, name, scope = parse_migration_filename(file)
         raise IllegalMigrationNameError.new(file) unless version
-        version = normalize_migration_number(version)
+        version = schema_migration.normalize_migration_number(version)
         status = db_list.delete(version) ? "up" : "down"
         [status, version, (name + scope).humanize]
       end.compact
```

Reply:
The entire class is nodoc'd.

Reply:
Took the majority of suggested changes and amended my original commit.